File: CS/ReportingBackend/Services/CustomReportResolver.cs
using System;
using System.Linq;
using System.Collections.Generic;
using DevExpress.XtraReports.UI;
using DevExpress.XtraReports.Web.WebDocumentViewer;

namespace ReportingBackend.Services {
    class ReportInfo {
        public string DisplayName { get; set; }
        public Func<XtraReport> GetReport { get; set; }
    }
    public class CustomReportResolver : IWebDocumentViewerReportResolver {
        Dictionary<string, ReportInfo> reports = new Dictionary<string, ReportInfo>();

        public CustomReportResolver() {
            reports.Add("ordersReport", new ReportInfo {
                DisplayName = "Orders Report",
                GetReport = () => new OrdersReport()
            });
            reports.Add("categoriesReport", new ReportInfo {
                DisplayName = "Categories Report",
                GetReport = () => new CategoriesReport()
            });
            reports.Add("shapeReport", new ReportInfo {
                DisplayName = "Shapes Report",
                GetReport = () => new ShapeReport()
            });
        }
        
        XtraReport IWebDocumentViewerReportResolver.Resolve(string reportEntry) {
            ReportInfo reportInfo;
            if(!string.IsNullOrEmpty(reportEntry) && reports.TryGetValue(reportEntry, out reportInfo))
                return reportInfo.GetReport.Invoke();
            return null;
        }

        // Get the list of available reports from the server.
        public Dictionary<string, string> GetAvailableReports() {
            return reports.ToDictionary(r => r.Key, r => r.Value.DisplayName);
        }


    }
}

File: VB/ReportingBackend/Services/CustomReportResolver.vb
Imports System
Imports System.Linq
Imports System.Collections.Generic
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraReports.Web.WebDocumentViewer

Namespace ReportingBackend.Services
    Friend Class ReportInfo
        Public Property DisplayName() As String
        Public Property GetReport() As Func(Of XtraReport)
    End Class
    Public Class CustomReportResolver
        Implements IWebDocumentViewerReportResolver

        Private reports As New Dictionary(Of String, ReportInfo)()

        Public Sub New()
            reports.Add("ordersReport", New ReportInfo With {.DisplayName = "Orders Report", .GetReport = Function() New OrdersReport()})
            reports.Add("categoriesReport", New ReportInfo With {.DisplayName = "Categories Report", .GetReport = Function() New CategoriesReport()})
            reports.Add("shapeReport", New ReportInfo With {.DisplayName = "Shapes Report", .GetReport = Function() New ShapeReport()})
        End Sub

        Private Function IWebDocumentViewerReportResolver_Resolve(ByVal reportEntry As String) As XtraReport Implements IWebDocumentViewerReportResolver.Resolve
            Dim reportInfo As ReportInfo = Nothing
            If (Not String.IsNullOrEmpty(reportEntry)) AndAlso reports.TryGetValue(reportEntry, reportInfo) Then
                Return reportInfo.GetReport.Invoke()
            End If
            Return Nothing
        End Function

        ' Get the list of available reports from the server.
        Public Function GetAvailableReports() As Dictionary(Of String, String)
            Return reports.ToDictionary(Function(r) r.Key, Function(r) r.Value.DisplayName)
        End Function


    End Class
End Namespace